docs: add modular design system adr (#8012)
primer/react: contributor-docs/adrs/adr-024-modular-ds.md

## Alternatives

The main alternatives to this approach would be:

- Do nothing, continue our current approach which is causing the challenges
  mentioned at the beginning of the ADR
- Fully adopt the layer model as proposed

Most likely we'll never find a model that will exactly work across components.
As a result, we should develop techniques and learn how to apply them as the
design system grows.

## Questions

### Will this be the de facto standard for all components in the Design System going forward?

Yes, our team will prioritize authoring components as presentational components
with hooks as the starting point for new components. As patterns and opinions
emerge, we create config components to simplify adoption of those patterns.

### Does this only apply to certain type of components?

Not every component will need to be authored with config, presentational, and
base components in mind. Not all components will need their own hooks, either.
Generally, when writing a component we will want to start with the
presentational components first. We'll focus on markup and styles before adding
in behavior and state. For some components, this can be baked into the

> component. For others, they will need to be extracted into hooks.

Reviewer: ❤️

# Modular Design System

📆 Date: 2026-05-17

## Status

| Stage          | State       |
| -------------- | ----------- |
| Status         | Proposed ⚠️ |
| Implementation | Pending ⚠️  |

## Context

There is a proposal for enabling better pattern creation via modular building
blocks in: https://github.com/github/core-ux/issues/2238. In this proposal, it
details several layers that Primer could be structured into:

- Layer 3: Ready-made, props-based components that are ready to use
- Layer 2: Parts, presentiontational-style components that one combines with hooks to
  create components
- Layer 1: Foundations, unstyled primitives to allow for full markup and style
  control
- Layer 0: Hooks, state management, and other logic that can be used to build components

We took this proposal into a Modular design System Workshop:
https://github.com/github/primer/discussions/6703. In this workshop, we

> identified several scenarios that team run into challenges when building with

Reviewer 2: ```suggestion
identified several scenarios where teams run into challenges when building with
```

Primer:

- It can be difficult to extend or add new functionality to a component
- It can be difficult to customize the appearance, content, or semantics of a
  component
- It can be difficult to build new experiences using established patterns (like accessibility primitives)

These challenges lead to custom implementations, forks, or overrides of Primer
components that reuslt in adoption challenges and ecosystem fragmentation.

As a result, we applied the layer model to several situations in a
[FigJam](https://www.figma.com/board/mgApQG7vL3imdUIhgyhG2n/Modular-DS-Workshop?node-id=84-152&t=dBKm8jx1UsKRXyzu-1). From this work, we pulled out insights about this model that we want to adopt in Primer today.

## Decision

From the Modular Design System workshop, we identified several key areas that we
want to focus on in order to improve Primer:

- Embrace layer 2 from the layer model, offering parts for our components that
  can be combined with hooks for improved flexibility
- Provide low-level components and hooks to make it easier to build

>   quality, accessible experience

Reviewer 2: ```suggestion
  quality, accessible experiences
```

> - Create a model for upstreaming work where people build on top of Primer

Reviewer 2: ```suggestion
- Create a model for upstreaming work where people have already built on top of Primer
```

To do this, we will:

> - Identify and deliver components to break down into parts that can be used to create more

Reviewer 2: ```suggestion
- Identify and deliver components that can be broken down into parts, which can be used to create more
```

  flexible and extensible components
- Identify and deliver low-level components, hooks, and utilities that can be used to quickly build out accessible experiences

> - Embrace the spectrum of abstraction model, authoring presentation components

Reviewer 2: ```suggestion
- Embrace the spectrum of abstraction model, authoring presentational components
```

  that combine with hooks to provide flexible patterns and opinionated defaults.

### Breaking components down into parts

We will break down the following components into parts plus appropriate
behavioral hooks. These changes will make it easier for teams to layer behavior
on top of these components.

- SelectPanel
- Dialog (and ConfirmationDialog)
- DataTable (write documentation for `useTable` and `Table.*` components)

### Delivering low-level components, hooks, and utilities

> We will immediately deliver high priority components that assist in building accessible experiences, including:

Reviewer 2: ```suggestion
We will immediately deliver high-priority components that will assist in building accessible experiences, including:
```

> - **Dialog** for modal and nonmodal experiences with built in parts for

Reviewer 2: ```suggestion
- **Dialog** for modal and nonmodal experiences with built-in parts for
```

> - **Dialog** for modal and nonmodal experiences with built in parts for

Reviewer 2: Is there something missing at the end of this line? 🤔

> - **ScrollableRegion** for scrollable containers with built in accessibility and behaviors

Reviewer 2: ```suggestion
- **ScrollableRegion** for scrollable containers with built-in accessibility and behaviors
```

|                |           | useResizeObserver            |
|                |           | useResponsiveValue           |
|                |           | useRovingTabIndex            |
|                |           | useSafeTimeout               |
|                |           | useSlots                     |
|                |           | useSyncedState               |
|                |           | useTheme                     |
| ./experimental | component | Hidden                       |
|                |           | Tabs                         |
|                | hook      | useSlots                     |
|                |           | useOverflow                  |
|                |           | useTab                       |
|                |           | useTabList                   |
|                |           | useTabPanel                  |

### Embracing the spectrum of abstraction model

We will author components with a spectrum of abstraction in mind, specifically
beginning with presentational components that can combine with hooks to provide
opionated, config-driven components.

This model allows us to offer opinionated defaults that work for common

> scenarios while still allow teams to extend and customize components as needed.

Reviewer 2: ```suggestion
scenarios while still allowing teams to extend and customize components as needed.
```

extend.

```tsx
<List
  items={[{label: 'Item one'}, {label: 'Item two'}, {label: 'Item three'}]}
  onSelect={item => {
    /* ... */
  }}
  onFilter={query => {
    /* ... */
  }}
/>
```

#### Presentational components

These components provide a lower level of abstraction and are more flexible and extensible, but require more work to build out common patterns and experiences.

```tsx
function Example({items}) {
  const [state, actions] = useList({
    defaultSelected: [],
  })

  return (
    <List>
      <List.Filter
        query={state.query}
        onChange={query => {
          actions.updateQuery(query)
        }}
      />
      {items
        .filter(item => item.label.includes(state.query))
        .map(item => {
          return (
            <List.Item
              key={item.label}
              onClick={() => {
                actions.toggleSelect(item.label)
              }}
            >
              <List.ItemLeadingVisual>
                <Checkbox checked={state.selected.has(item.label)} />
              </List.ItemLeadingVisual>
              <List.ItemLabel>{item.label}</List.ItemLabel>
            </List.Item>
          )
        })}
    </List>
  )
}
```

> Presentational components are accompanied by behavior or state hooks for a
> component's core functionality. Often times, config components are built by
> combining presentational components and the corresponding behavior and state
> hooks needed for a feature.

Reviewer 2: ```suggestion
Presentational components are accompanied by behavior or state hooks for a
component's core functionality. Oftentimes, config components are built by
combining presentational components with the corresponding behavior and state
hooks needed for a feature.
```

# Modular Design System

📆 Date: 2026-05-17

## Status

| Stage          | State       |
| -------------- | ----------- |
| Status         | Proposed ⚠️ |
| Implementation | Pending ⚠️  |

## Context

There is a proposal for enabling better pattern creation via modular building
blocks in: https://github.com/github/core-ux/issues/2238. In this proposal, it
details several layers that Primer could be structured into:

- Layer 3: Ready-made, props-based components that are ready to use

> - Layer 2: Parts, presentiontational-style components that one combines with hooks to

Reviewer 3: ```suggestion
- Layer 2: Parts, presentational-style components that one combines with hooks to
```

  create components
- Layer 1: Foundations, unstyled primitives to allow for full markup and style
  control
- Layer 0: Hooks, state management, and other logic that can be used to build components

> We took this proposal into a Modular design System Workshop:

Reviewer 3: ```suggestion
We took this proposal into a Modular Design System Workshop:
```

https://github.com/github/primer/discussions/6703. In this workshop, we
identified several scenarios where teams run into challenges when building with
Primer:

- It can be difficult to extend or add new functionality to a component
- It can be difficult to customize the appearance, content, or semantics of a
  component
- It can be difficult to build new experiences using established patterns (like accessibility primitives)

These challenges lead to custom implementations, forks, or overrides of Primer

> components that reuslt in adoption challenges and ecosystem fragmentation.

Reviewer 3: ```suggestion
components that result in adoption challenges and ecosystem fragmentation.
```

As a result, we applied the layer model to several situations in a
[FigJam](https://www.figma.com/board/mgApQG7vL3imdUIhgyhG2n/Modular-DS-Workshop?node-id=84-152&t=dBKm8jx1UsKRXyzu-1). From this work, we pulled out insights about this model that we want to adopt in Primer today.

## Decision

From the Modular Design System workshop, we identified several key areas that we
want to focus on in order to improve Primer:

- Embrace layer 2 from the layer model, offering parts for our components that
  can be combined with hooks for improved flexibility

> - Provide low-level components and hooks to make it easier to build

Reviewer 3: ```suggestion
- Provide low-level components and hooks to make it easier to build customized, 
```

Reviewer 3: Correct me if this isn't quite right, but the way I've been thinking about _why_ we want to offer these is truly for cases where the consumer wants more customization than the ready-made component offers. I think calling that out explicitly may be helpful.

  quality, accessible experiences
- Create a model for upstreaming work where people have already built on top of Primer

To do this, we will:

- Identify and deliver components that can be broken down into parts, which can be used to create more
  flexible and extensible components
- Identify and deliver low-level components, hooks, and utilities that can be used to quickly build out accessible experiences
- Embrace the spectrum of abstraction model, authoring presentational components
  that combine with hooks to provide flexible patterns and opinionated defaults.

### Breaking components down into parts

We will break down the following components into parts plus appropriate
behavioral hooks. These changes will make it easier for teams to layer behavior
on top of these components.

- SelectPanel
- Dialog (and ConfirmationDialog)
- DataTable (write documentation for `useTable` and `Table.*` components)

> - ActionList (break down into list primitives)

Reviewer 3: Once there's alignment, a helpful next step from this ADR would be to estimate the staffing and time we'd need to break each of these components into the layer 2 parts with any necessary low-level hooks/utilities 😄

### Delivering low-level components, hooks, and utilities

> We will immediately deliver high-priority components that will assist in building accessible experiences, including:

Reviewer 3: ```suggestion
We will [scope and plan](https://github.com/github/primer/issues/6773) to deliver high-priority components that will assist in building accessible experiences, including:
```

- **Dialog** for modal and nonmodal experiences with built-in parts for dialog
  headings, content, and actions
- **ScrollableRegion** for scrollable containers with built-in accessibility and behaviors
- **Popover** for positioned, transient experiences
- **Tabs** for building tabs-based experiences
- **ToggleButton**, **ToggleSwitch** for building toggle-based experiences
- **Disclosure** for showing/hiding content
- **Accordion** for showing/hiding content

We will also deliver documentation for all existing low-level components,
hooks, and utilities.

In the future, we are investigating the following opportunities for offering
additional low-level components, hooks, and utilities:

- **Listbox** for selection-based experiences
- **Combobox** for filtering and selection-based experiences
  wiring up accessibility, behaviors, and semantics
- **Menu** for action-based experiences
- **Tree** for tree-based experiences
- useFilter
- useSelection

#### Audit

| Entrypoint     | Type      | Name                         |

> | :------------- | :-------- | :--------------------------- |

Reviewer: I don't understand what this Audit portion is and how it ties into this ADR 😅

Author: Basically with the different layers or areas being composed, the audits were showing what work was left to accomplish them. They're removed now though!

|                |           | useOnOutsideClick            |
|                |           | useOpenAndCloseFocus         |
|                |           | useOverlay                   |
|                |           | useProvidedRefOrCreate       |
|                |           | useRefObjectAsForwardedRef   |
|                |           | useResizeObserver            |
|                |           | useResponsiveValue           |
|                |           | useRovingTabIndex            |
|                |           | useSafeTimeout               |
|                |           | useSlots                     |
|                |           | useSyncedState               |
|                |           | useTheme                     |
| ./experimental | component | Hidden                       |
|                |           | Tabs                         |
|                | hook      | useSlots                     |
|                |           | useOverflow                  |
|                |           | useTab                       |
|                |           | useTabList                   |
|                |           | useTabPanel                  |

> ### Embracing the spectrum of abstraction model

Reviewer 3: I'd love to embrace this from a future-looking perspective by picking up https://github.com/github/primer/issues/6789 before we close out the current scope of modular DS investigation.

We will author components with a spectrum of abstraction in mind, specifically
beginning with presentational components that can combine with hooks to provide

> opionated, config-driven components.

Reviewer 3: ```suggestion
opinionated, config-driven components.
```

    defaultSelected: [],
  })

  return (
    <List>
      <List.Filter
        query={state.query}
        onChange={query => {
          actions.updateQuery(query)
        }}
      />
      {items
        .filter(item => item.label.includes(state.query))
        .map(item => {
          return (
            <List.Item
              key={item.label}
              onClick={() => {
                actions.toggleSelect(item.label)
              }}
            >
              <List.ItemLeadingVisual>
                <Checkbox checked={state.selected.has(item.label)} />
              </List.ItemLeadingVisual>
              <List.ItemLabel>{item.label}</List.ItemLabel>
            </List.Item>
          )
        })}
    </List>
  )
}
```

Presentational components are accompanied by behavior or state hooks for a
component's core functionality. Oftentimes, config components are built by
combining presentational components with the corresponding behavior and state
hooks needed for a feature.

#### Base components

Base components are unstyled and may optionally provide behavior. These
components are fundamental primitives that are used to build components, such as

> accessibility primitives.

Reviewer: nit: I think its worth filling out a bit more info here. I was catching up with the last eng session we had (06/25) where you (Josh) were explaining the use case for these (as I understood it pseudoinheritance: two somewhat similar components can consume the same Bases and it also enables consumer to build entirely custom new ones). One thing that isn't clear to me from this is whether the aim is we'll have Base components for all components we "migrate" into the layered model, or do we expect only those that have this "need" to go down to this layer? 
May also be a good part of the #questions section: when does a component implement Layer 0/1/2/3?

Author: @francinelucca as far as I'm aware, this would just be for things we want to share between our components. For example, two components should not have to reimplement listbox or tab semantics. Instead, there should be a "base" version of tabs that is shared across our components.

The reason it comes up as something we want to export from PRC is that other teams also have this desire to build on top of tabs, listbox semantics, etc

```tsx
function Example() {
  return (
    <Dialog.Root>
      <Dialog.Trigger />
      <Dialog.Overlay />
      <Dialog.Content>
        <Dialog.Close />
      </Dialog.Content>
    </Dialog.Root>
  )
}
```

Other examples include:

- Combobox (filtering, selection)
- Listbox (selection)
- Popover
- Tabs
- Treeview

#### Utilities

There are core utilities, hooks, functions, etc that are commonly used when
building out components or React applications. We provide hooks for established
patterns so that teams can build on solid foundations when authoring new
experiences. These can include hooks such as `useMergedRefs`,
`useOnEscapePress`, `useTimeout` and more.

These utilities may extend beyond hooks, such as `@primer/behaviors` or custom
elements, where appropriate.

## Consequences

## Alternatives

> ## Questions

Reviewer 3: Are any of these sections worth filling in? If so, maybe Copilot could help author quickly? For instance, one alternative is to leave most components as "ready-made" though the consequence of that is that we aren't providing deeper levels of customization to consumers (like design engineers) who are asking for it.

Reviewer: Or to slot-ify the entire library. but I feel like that also is too rigid for the level of customization we're trying to enable

Author: Added some blurbs to this! Feel free to suggest any other ones and I'd be happy to include them 😄

> ## Questions

Reviewer: - Will modular be the de facto standard for all components in the Design System going forward? or does it only apply to certain type of components?
- Do all components that adopt the modular strategy implement every layer, or do different component have different "needs"? how is this decided? 

("These questions will be answered by https://github.com/github/primer/issues/6789 or xyz tracking link" is a perfectly acceptable answer)
